File: WaveWriter.m
classdef WaveWriter < Module
   
    properties(SetAccess = private)
   
        leftInput
        rightInput
        y
        
    end
   
    methods
       
        function this = WaveWriter(name)
           
            this = this@Module(name);
            
            this.leftInput = this.createInputPort();
            this.rightInput = this.createInputPort();
            
        end
        
        function doUpdate(this, N)
            
            this.y = [this.leftInput.read(N) this.rightInput.read(N)];
            
        end;
    end
end

File: Cpp/Modules/Source/WaveWriter.cpp
#include "WaveWriter.h"


SoundHeader& SetupHeader(SoundHeader& header, int rate)
{
	header.setHighStereo();
	header.setSrate(rate);
	return header;
}

WaveWriter::WaveWriter(const char* fileName, int rate)
	: _pLeftInput(CreateInputPort())
	,_pRightInput(CreateInputPort())
	,_soundFileWrite(fileName, SetupHeader(_header, rate))
{
}

void WaveWriter::Update()
{
	_soundFileWrite.writeSampleDouble(_pLeftInput->Read());
	_soundFileWrite.writeSampleDouble(_pRightInput->Read());
}

void WaveWriter::Close()
{
	_soundFileWrite.close();
}

IInputPort* WaveWriter::LeftInput() const
{
	return _pLeftInput;
}

IInputPort* WaveWriter::RightInput() const
{
	return _pRightInput;
}
